### ISSUE TITLE
Resize operation fails (['unk__0', 'unk__1', 'unk__2', 'unk__3']) and raises UnboundLocalError: local variable 'new_size' referenced before assignment

### ISSUE THREAD

**Issue by CarlosNacher** [PINTO0309/onnx2tf#607]
### Issue Type

Others

### OS

Linux

### onnx2tf version number

1.19.15

### onnx version number

1.15.0

### onnxruntime version number

Version: 1.17.1

### onnxsim (onnx_simplifier) version number

0.4.33

### tensorflow version number

2.15.0.post1

### Download URL for ONNX

https://we.tl/t-HJjXDINMND

### Parameter Replacement JSON

```json
{
  "format_version": 1,
  "operations": [
    {
      "op_name": "/ae/decoder/Resize",
      "param_target": "outputs", 
      "param_name": "/ae/decoder/Resize_output_0",
      "values": [1, 64, 25, 12] 
    }
  ]
}
```

### Description

Link to the Google Colab notebook with all the execution: https://colab.research.google.com/drive/1StDa10u2DytLO_8IivYMMUNuI_dndLpr?usp=sharing

1. Purpose: **I am trying to convert `.onnx` model into `.pb`** one because I want to embed my model in a Google Coral TPU USB Accelerator. So following: https://coral.ai/docs/edgetpu/models-intro/#compatibility-overview, the first I have to do is convert my pytorch model into onnx (done), then to tensorflow (i am stuck with it), then quantize, convert to tfilte, etc, etc
2. What: **My model seems to be failing in a Resize operation, I am providing param_replacement JSON, but I don't know if I am not using it properly, because it seems to have no effect.**
3. How: Yo can find the link to the entire execution pipeline in the Google Colab link at the init of this comment.
4. Why: because if not, I can not continue with my purpose of embedding the model into a TPU-
5. Resources: No additional of this repo.

Please, I am struggling a lot with this. Your tool is the most hopeful I have found after a intense research, however I don't think I fully understand how to use it and get the most out of it. If you can help me, I would really appreciate it!

Thank you so much in advance!

PS: I re-write the links of my `.onnx` model: https://we.tl/t-HJjXDINMND and my colab notebook: https://colab.research.google.com/drive/1StDa10u2DytLO_8IivYMMUNuI_dndLpr?usp=sharing

**Comment by CarlosNacher** [comment]
**Update:** I think I have solved it with the `-ois` parameter. At first, I was trying `!onnx2tf -i $PATH_TO_MY_MODEL -ois data:1,3,1024,608` because "data" is what is said to be used in the README, but I have just realized that it has to be the name of the onnx input, in my case "input", so with `!onnx2tf -i $PATH_TO_MY_MODEL -ois input:1,3,1024,608` it works.

However, I still doesn`t fully understand the outputs, since I don't know wether the following is a problem that I have to try to solve and how to do it:
```
Summary on the non-converted ops:
---------------------------------
 * Accepted dialects: tfl, builtin, func
 * Non-Converted Ops: 68, Total Ops 138, % non-converted = 49.28 %
 * 68 ARITH ops

- arith.constant:   68 occurrences  (f32: 52, i32: 16)
```

**Comment by PINTO0309** [comment]
Your last command you tried is correct. There is nothing we engineers can do about that log that TensorFlow displays for no good reason.

```
onnx2tf -i model.onnx -cotof -ois input:1,3,1024,608
```

The outputs of ONNX and TFLite agree with each other with an error of less than `1e-4`.

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/5d08fddb-33ff-450f-82fe-fd429ced8e5f)

**Comment by CarlosNacher** [comment]
> Your last command you tried is correct. There is nothing we engineers can do about that log that TensorFlow displays for no good reason.
> 
> ```
> onnx2tf -i model.onnx -cotof -ois input:1,3,1024,608
> ```
> 
> The outputs of ONNX and TFLite agree with each other with an error of less than `1e-4`.
> 
> ![image](https://private-user-images.githubusercontent.com/33194443/322802046-5d08fddb-33ff-450f-82fe-fd429ced8e5f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyNjk3ODQsIm5iZiI6MTcxMzI2OTQ4NCwicGF0aCI6Ii8zMzE5NDQ0My8zMjI4MDIwNDYtNWQwOGZkZGItMzNmZi00NTBmLTgyZmUtZmQ0MjljZWQ4ZTVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE2VDEyMTEyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiNTNiOTczNjFmODVlNGJhMDBhYmZhZGQ3MDIxYTVlN2ZiNzk5MDRjYjE5OTVkMTI4ZGI4NmY4ZjcwZGJkZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.dHs-yEXTXLusXzjBXRENM_dTfiMrvrweBUhwAiVLLoc)
> 
> By the way, your model cannot be converted to a model for edgetpu.

Hi, thanks for your quick response! So, is it usual to have some operations "not converted"? And, why my model cannot be converted to a model for edgetpu? :/

**Comment by PINTO0309** [comment]
>So, is it usual to have some operations "not converted"? 

I don't know. I don't even know what "not converted" means. Ask for help on the TensorFlow forum. The internal specifications of TensorFlow are outside the scope of this repository.

> why my model cannot be converted to a model for edgetpu?

 Simply, the input resolution is too large.  Check the TensorFlow and EdgeTPU documentation yourself.

**Comment by CarlosNacher** [comment]
> > So, is it usual to have some operations "not converted"?
> 
> I don't know. I don't even know what "not converted" means. Ask for help on the TensorFlow forum. The internal specifications of TensorFlow are outside the scope of this repository.
> 
> > why my model cannot be converted to a model for edgetpu?
> 
> Simply, the input resolution is too large. Check the TensorFlow and EdgeTPU documentation yourself.

The Edge TPU documentation about model requirements (https://coral.ai/docs/edgetpu/models-intro/#model-requirements) don't say anything about input resolution. It only restricts all operations to be INT8, to be also in the list of supported operations and it is supposed that have to weigh less than 2GB (the memory of the TPU). Can you share me the part of the documentation where they talk about the input resolution, please?

**Comment by PINTO0309** [comment]
>Can you share me the part of the documentation where they talk about the input resolution, please?

 I have no doubt that I have undocumented knowledge, including what is documented. I've been working on TensorFlow and EdgeTPU far longer than you all have, and I've come to that answer by actually transforming models and failing, and by looking all over the issues.

Quite simply, Google engineers do not attempt to answer questions seriously. For specifications not listed in their documentation, there is no other way but to read the source code of the TensorFlow and EdgeTPU runtimes. However, I do not remember or track the location of every single relevant source code. That's because the source code has changed significantly over the years.

**Comment by CarlosNacher** [comment]
> Quite simply, Google engineers do not attempt to answer questions seriously. For specifications not listed in their documentation, there is no other way but to read the source code of the TensorFlow and EdgeTPU runtimes. However, I do not remember or track the location of every single relevant source code. That's because the source code has changed significantly over the years.

Okay, so the point is that you think it will fail based on your expertise, right? I respect your expertise since you have done a great job with this library, I am not calling into question your response, only I want to understand why model could fail, if an 3x1024x608 image weighs less than 2Mb and the memory of the TPU is 2GB. I only want to understand you, since I am interested in making it work.

Thank you!

**Comment by CarlosNacher** [comment]
Yes, I don't know why Google doesn't provide good documentation and code for TPU usage.

I have tried the command `!onnx2tf -i $PATH_TO_MY_MODEL -ois input:1,3,1024,608 -coto -oiqt`

But the actual error is not related to resolution:
```
...
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/lite/python/optimize/calibrator.py", line 101, in _feed_tensors
    for sample in dataset_gen():
  File "/usr/local/lib/python3.10/dist-packages/onnx2tf/onnx2tf.py", line 1435, in representative_dataset_gen
    yield_data_dict[model_input_name] = normalized_calib_data.astype(np.float32)
  File "/usr/local/lib/python3.10/dist-packages/tensorflow/python/framework/tensor.py", line 256, in __getattr__

AttributeError: EagerTensor object has no attribute 'astype'. 
        If you are looking for numpy-related methods, please run the following:
        tf.experimental.numpy.experimental_enable_numpy_behavior()
      . Did you mean: 'dtype'?
```

**Comment by PINTO0309** [comment]
I don't understand why such an error would occur even though it is a numpy.ndarray variable. Try downgrading to `tensorflow==2.14.0`.

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/a14bf1ac-9962-4f83-98ea-6cf39d676812)

**Comment by PINTO0309** [comment]
lol. I reproduced it. Obviously, this is a bug in TensorFlow. Wait a few hours to add the workaround fix to onnx2tf.

![image](https://github.com/PINTO0309/onnx2tf/assets/33194443/68c471d3-f850-4b08-9c65-82ccd54d0c34)

**Comment by CarlosNacher** [comment]
Okay, thank you so much!

**Comment by PINTO0309** [comment]
They are a super talented group of engineers, but they have no interest in documentation.

**Comment by CarlosNacher** [comment]
Yes but Google (not only engineers, the company in general) wants to earn money, and if people doesn't know how to use its products, they won't use it, and they won't buy it. The only reason comes t my mind is they prefer the GCP bussiness, idk

**Comment by PINTO0309** [comment]
```
onnx2tf -i model.onnx -oiqt -qt per-tensor -ois input:1,3,1024,608
```
[model_full_integer_quant.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/14998270/model_full_integer_quant.tflite.zip)

**Comment by PINTO0309** [comment]
Fixes: https://github.com/PINTO0309/onnx2tf/releases/tag/1.19.16

**Comment by PINTO0309** [comment]
This question has nothing to do with onnx2tf, it's basic quantization. Quantization using the -cind option of onnx2tf and quantization directly using TFLiteConverter are the same operation. I am not your teacher and I am not going to answer any more essentially irrelevant questions.

The normalization method and its scope will vary depending on whether your model entrance includes a normalization layer or not, and on the content of the input data. If a normalization layer is included in the model, mean=0.0 and std=1.0 should be used.

https://www.tensorflow.org/lite/performance/post_training_quantization

https://www.tensorflow.org/lite/performance/quantization_spec

Inside onnx2tf, the tensor entered with the -cind option is not divided by 255.0. Therefore, it is only necessary to divide all data for calibration by 255.0 in advance. However, this is the case for RGB input data in the range 0-255, and for non-image data, division by 255 is wrong.

**Comment by CarlosNacher** [comment]
I know, the only I want to know is, in the fully INT quantized model, how to calibrate future inference data in the same way is was done during the model exporting. But if you are not my teacher it's okay, one love!
